formal-spec/measurements.tex

```diff
@@ -174,6 +177,102 @@ \section{Impact on performance}
+tallyStake confidence result ballot stakeDistribution adversarialStakeRatio =
+  if stakeThreshold adversarialStakeRatio (totalStake stakeDistribution)
+     <
+     stakeOfKeys votingKeys stakeDistribution
+  then Just result
+  else Nothing
+  where
+    votingKeys = Map.filter (== confidence) ballot
+\end{lstlisting}
+
+\lstinline{Map.filter} is $O(n)$ \footnote{\href{url}{http://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Strict.html\#g:25}}, so \lstinline{votingKeys} is $O(n)$, where $n$ is the number of ballots, which as we have said coincides to the number of participants. At this point, we have a single pass (loop) over $n$ ballots.
+
+Furthermore, \lstinline{stakeOfKeys} makes the following calls:
+\begin{lstlisting}[language=Haskell, caption=Code example]
+stakeOfKeys
+  keyMap
+  StakeDistribution
+  { stakeMap
+  }
+  = Map.foldl (+) 0 $ stakeMap `Map.intersection` keyMap
+\end{lstlisting}
+
+The $intersection$ function in the worst-case is $O(n)$\footnote{\href{url}{http://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Strict.html\#v:intersection}}. Therefore this is a second pass (loop) over the data of length $n$.
+\lstinline{foldl} is also $O(n)$\footnote{\href{url}{http://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Strict.html\#v:foldl}}. This is a third pass (loop) over the data of length $n$.
+Thus from the above analysis we see that we have for a single proposal three calls of \lstinline{tallyStake}, where in each one we have three passes over the data of length $n$. So in total for a single proposal we do $9$ passes over the data of length $n$. That is $9n$ operations, which means that the tally time complexity is $O(n)$. 
+
+This result is also confirmed by the experimental evaluation shown in the graph below:
+
+\begin{figure}[htp]
+  \centering
+
+  \begin{tikzpicture}
+    \begin{axis}[
+      title={Elapsed time (sec) vs Number of participants},
+      xlabel={Participants},
+      xmin=100.0,
+      xmax=1.0e7,
+      xmode=log,
+      xtick={10.0, 100.0, 1000.0, 10000.0, 100000.0, 1000000.0, 1.0e7},
+      ylabel={Elapsed time (sec)},
+      ymin=0.0000001,
+      ymax=10,
+      ymode=log, 
+      ytick={0.0000001, 0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1, 1, 10},
+      legend pos=north west,
+      ymajorgrids=true,
+      grid style=dashed,
+      ]
+      \addplot[color=black] table {participants-vs-elapsed_time.dat};
+    \end{axis}
+  \end{tikzpicture}
+
+  \caption{Worst case scenario analysis for tally phase processing time}
+  \label{fig:eltime-vs-participants}
+\end{figure}
+
+In Figure \ref{fig:eltime-vs-participants}, we see the elapsed time of the tally phase to scale linearly in the number of participants. In addition, we see that it takes almost one tenth of a second to process the votes of $1$ million participants. These results correspond to a no-parallel execution of the tally algorithm on a i7 CPU laptop with 32GB of RAM.
```

Comment:
I'd say "the elapsed time **increases** linearly"

Reply:
OK